### ISSUE TITLE
Document limitations when using lazy loaded dependencies that involve variadic functions

### ISSUE THREAD

**Issue by holtkamp** [PHP-DI/PHP-DI#359]
When using lazy loaded dependencies, I encountered some curious behaviour when handing over more parameters to a function than defined in its signature ([variadic functions](https://en.wikipedia.org/wiki/Variadic_function)). 

```php
class Service
{
	/**
	 * @param mixed $param1
	 * @param mixed $param2
	 * @param mixed $paramN Can be used to supply additional parameters
	 */
	public function doStuf($param1, $param2){
		echo func_num_args();
	}
}
```

Now let's try to use this Service when fetched from the dependency injection container

```php
$dic->get('Service')->doStuff(1, 2, 3); //Prints '3' when not lazy loaded
$dic->get('Service')->doStuff(1, 2, 3); //Prints '2' when lazy loaded
```
I use this approach to inject extra variables when translating messages like: "User X has logged in on Y", where X and Y would be 'additional' parameters.

This probably has to do with the intermediate proxy classes that are generated. For example from a generated proxy class:

```php
  /**
     * {@inheritDoc}
     */
    public function doStuf($param1, $param2)
    {
        $this->initializer5683c4a86687a054344119 && $this->initializer5683c4a86687a054344119->__invoke($this->valueHolder5683c4a866860700867865, $this, 'fireEventOnOrder', array('param1' => $param1, 'param2' => $param2), $this->initializer5683c4a86687a054344119);

        return $this->valueHolder5683c4a866860700867865->fireEventOnOrder($param1, $param2);
    }
```

When looking at https://github.com/Ocramius/ProxyManager/issues/177, I think this behaviour is expected. Therefore I think it should be mentioned in the documentation as a warning...

**Comment by mnapoli** [comment]
Well thank you for going so far in the debugging :)

Ping @Ocramius. It seems https://github.com/Ocramius/ProxyManager/issues/177 is about supporting PHP 5.6's variadics (the `...` syntax), so I'm guessing it might simply be a bug in ProxyManager? (i.e. it should use `func_get_args()` instead of listing each parameter?)

**Comment by Ocramius** [comment]
@mnapoli I still need to pack variadics, scalar type hints and return type hints into ProxyManager.

**Comment by mnapoli** [comment]
Just to be sure, when you say variadics you also mean "PHP < 5.6 kind of variadics" which consists of passing more parameters than defined right?

In that case indeed we could document in PHP-DI that this is not supported right now to avoid confusion.

**Comment by Ocramius** [comment]
> Just to be sure, when you say variadics you also mean "PHP < 5.6 kind of variadics" which consists of passing more parameters than defined right?

No, I mean 5.6 variadics (`...` operator).

Pre-PHP-5.6 "variadic arguments" are totally ignored. Going to open an issue for that now.

**Comment by Ocramius** [comment]
Note: this is also broken in doctrine2, now that I notice it (known, but undocumented limitation).

Actually, the lazy loading ghost objects in ProxyManager 2.x work with that one, but I still need to implement them.

Opened Ocramius/ProxyManager#265 to track that, thanks!

**Comment by mnapoli** [comment]
Thanks!

**Comment by Ocramius** [comment]
I provided documentation for this sort of problem [in ProxyManager](https://github.com/Ocramius/ProxyManager/blob/450739b8d4c5661e7314f9e431bc04eac7d494e7/docs/lazy-loading-value-holder.md#known-limitations) and fixed the support of variadic arguments in the 2.x branch.

There is also a possible resolution path for this particular issue at Ocramius/ProxyManager#269, but it is complex and for a super-edge case.

I'll get a 2.0 release out soon, but be aware that:

 * it will work with PHP "~5.6|~7.0"
 * there will be some BC breaks

**Comment by mnapoli** [comment]
:+1: I was just reading this ;) Thanks

**Comment by holtkamp** [comment]
just checking on the status on this to see whether this issue can be closed...

[ProxyManager 2.0 have been released for a while now](https://github.com/Ocramius/ProxyManager/releases) and https://github.com/Ocramius/ProxyManager/issues/265 is closed. But I am not sure whether the named 'Ghost objects' which should support the pre-PHP-5.6 variadic functions is already implemented.

**Comment by mnapoli** [comment]
But it seems that v2 requires PHP 7 right?

**Comment by holtkamp** [comment]
aah, yeah seems to be the case: https://github.com/Ocramius/ProxyManager/blob/6c89b7bd6039d8047b1473e2074cb56baa4bc15d/composer.json#L22

Can we then conclude this will not be resolved for the time being and this limitation needs to be documented?

**Comment by mnapoli** [comment]
Yep!

**Comment by Ocramius** [comment]
Note: you can require `~1.0|~2.0`, since the API didn't change massively. Please see https://github.com/symfony/symfony/pull/17919 for a reference implementation of a BC compliant way of doing this.

As for the rest: you simply need to document that lazy services with variadic signatures require PHP 7 and the latest ProxyManager.

**Comment by mnapoli** [comment]
:+1: good idea! Thanks for the links

**Comment by holtkamp** [comment]
Aah, nice approach :wink:

**Comment by holtkamp** [comment]
@mnapoli are you considering adopting the suggestion of @Ocramius ? Guess it would be similar to: https://github.com/doctrine/migrations/blob/5f9429e5e6dd31d0f36c8951daa184112a39e195/composer.json#L18

**Comment by mnapoli** [comment]
Oh yes I'm definitely up for it, I just didn't have time to implement it yet. I guess what's to do:

- update [ProxyFactory](https://github.com/PHP-DI/PHP-DI/blob/master/src/DI/Proxy/ProxyFactory.php) if necessary to make it compatible with both version (since it's very simple it's possible there is nothing to do)
- require `~1.0|~2.0`, tests should run with highest and lowest version on Travis automatically
- update the documentation

If you are up for it I'll be :+1: to merge it!

**Comment by mnapoli** [comment]
Fixed in #414 (PHP-DI can now be installed with ProxyManager ~1.0 or ~2.0).

**Comment by Ocramius** [comment]
@mnapoli awesome! I really hope to push those 2.0 installations up (too many 1.x users ;-) )
